backends/ebpf/psa/ebpfPsaArch.cpp

```diff
@@ -0,0 +1,637 @@
+    control->headers = *it; ++it;
+    control->user_metadata = *it; ++it;
+    control->inputStandardMetadata = *it; ++it;
+    control->outputStandardMetadata = *it;
+
+    auto codegen = new ControlBodyTranslator(control);
+    codegen->substitute(control->headers, parserHeaders);
+
+    if (type != TC_EGRESS) {
+        codegen->useAsPointerVariable(control->outputStandardMetadata->name.name);
+    }
+
+    codegen->useAsPointerVariable(control->headers->name.name);
+    codegen->useAsPointerVariable(control->user_metadata->name.name);
+
+    control->codeGen = codegen;
+
+    for (auto a : ctrl->constantValue) {
+        auto b = a.second;
+        if (b->is<IR::Block>()) {
+            this->visit(b->to<IR::Block>());
+        }
+    }
+    return true;
+}
+
+bool ConvertToEBPFControlPSA::preorder(const IR::TableBlock *tblblk) {
+    // use HASH_MAP as default type
+    TableKind tableKind = TableHash;
+
+    // If any key field is LPM we will generate an LPM table
+    auto keyGenerator = tblblk->container->getKey();
+    if (keyGenerator != nullptr) {
+        for (auto it : keyGenerator->keyElements) {
+            // optimization: check if we should generate timestamp
+            if (it->expression->toString().endsWith("timestamp")) {
+                control->timestampIsUsed = true;
+            }
+
+            auto mtdecl = refmap->getDeclaration(it->matchType->path, true);
+            auto matchType = mtdecl->getNode()->to<IR::Declaration_ID>();
+            if (matchType->name.name != P4::P4CoreLibrary::instance.exactMatch.name &&
+                matchType->name.name != P4::P4CoreLibrary::instance.lpmMatch.name)
+                ::error(ErrorType::ERR_UNSUPPORTED,
+                        "Match of type %1% not supported", it->matchType);
+
+            if (matchType->name.name == P4::P4CoreLibrary::instance.lpmMatch.name) {
+                if (tableKind == TableLPMTrie) {
+                    ::error(ErrorType::ERR_UNSUPPORTED,
+                            "%1%: only one LPM field allowed", it->matchType);
+                    return false;
+                }
+                tableKind = TableLPMTrie;
+            }
+        }
+    }
+
+    // use 1024 by default
+    size_t size = 1024;
+    auto sizeProperty = tblblk->container->properties->getProperty(
+            IR::TableProperties::sizePropertyName);
+    if (sizeProperty != nullptr) {
+        auto expr = sizeProperty->value->to<IR::ExpressionValue>()->expression;
+        size = expr->to<IR::Constant>()->asInt();
+    }
+
+    cstring name = EBPFObject::externalName(tblblk->container);
+
+    EBPFTablePSA *table = new EBPFTablePSA(program, tblblk, control->codeGen, name, size);
+
+    control->tables.emplace(tblblk->container->name, table);
+    return true;
+}
+
+bool ConvertToEBPFControlPSA::preorder(const IR::AssignmentStatement *a) {
+    // the condition covers both ingress and egress timestamp
+    if (a->right->toString().endsWith("timestamp")) {
```

Comment:
This is somewhat ugly.

Reply:
Yeah, I agree it is. The goal is to find out if `ingress_timestamp` or `egress_timestamp` is used in a control block. We haven't found a better way to do this so far.